### PR TITLE
Add NET 8.0 target to Microsoft.Bcl.Memory package

### LINKED ISSUE #104185: Add NET 8.0 target to Microsoft.Bcl.Memory package
### Description

.NET 9.0 introduced `Base64Url`, and we are providing a polyfill for this type in `Microsoft.Bcl.Memory` to enable its use in earlier versions of .NET. The `Base64Url` implementation in .NET 9.0 leverages vectorization for improved performance. However, the polyfill implementation does not use vectorization, as it is not available in older framework versions. Although vectorization is possible in .NET 8.0, `Microsoft.Bcl.Memory` does not currently target .NET 8.0 specifically. When using `Microsoft.Bcl.Memory` on .NET 8.0, the polyfill implementation without vectorization is used. This issue tracks the addition of a .NET 8.0 target to `Microsoft.Bcl.Memory`, which will enable vectorization as well.

A quick test indicates that the current implementation, which uses vectorization, relies on the internal members of the public `Ascii` class: `VectorContainsNonAsciiChar`, `ExtractAsciiVector`, `StoreVectorAndZip`, `Load4xVector128AndUnzip`, and `ShuffleUnsafe`. These members need to be refactored to enable their use when targeting .NET 8.0.

### Reproduction Steps

Run any test app on .NET 8.0 with referencing `Microsoft.Bcl.Memory`.

### Expected behavior

Vectorization should be used

### Actual behavior

No Vectorization is used

### Regression?

No

### Known Workarounds

Target .NET 9.0 instead.

### Configuration

.NET 8.0

### Other information

_No response_

### PR DESCRIPTION
**Pull request by buyaa-n**
- #if-def-ed some new APIs that newly added in .NET 9.0 like `CompExactlyDependsOn` attribute
- Added .NET 8.0 copies for internal APIs that exist in 9.0 but not in 8.0 like: `Ascii.VectorContainsNonAsciiChar`, `Ascii.ExtractAsciiVector`, `Vector128.SuffleUnsafe`. 
- Some APIs in ARM vectorization path not exist in 8.0, so made the ARM code path only for .NET 9

Fixes https://github.com/dotnet/runtime/issues/104185

### CODE REVIEW

File: src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj [hunk 1]
```diff
@@ -1,14 +1,14 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;netstandard2.1;$(NetFrameworkMinimum);$(NetCoreAppCurrent)</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;netstandard2.1;$(NetFrameworkMinimum);$(NetCoreAppMinimum);$(NetCoreAppCurrent)</TargetFrameworks>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp' or $([MSBuild]::VersionLessThan('$(TargetFrameworkVersion)', '9.0'))">
```
Comment (tarekgh) [inline src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:7]: >"'$(TargetFrameworkIdentifier)' != '.NETCoreApp' or $([MSBuild]::VersionLessThan('$(TargetFrameworkVersion)', '9.0'))"> [](http://example.com/codeflow?start=23&length=119)

!IsTargetFrameworkCompatible('$(TargetFramework)', 'net8.0')

File: src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
     <Compile Include="Microsoft.Bcl.Memory.cs" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '9.0'))">
```
Comment (tarekgh) [inline src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:11]: >'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '9.0')) [](http://example.com/codeflow?start=24&length=128)

IsTargetFrameworkCompatible('$(TargetFramework)', 'net8.0')

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj [hunk 3]
```diff
@@ -34,7 +34,7 @@
     <Compile Include="..\ref\Microsoft.Bcl.Memory.Forwards.netstandard21.cs" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp' or $([MSBuild]::VersionLessThan('$(TargetFrameworkVersion)', '9.0'))">
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:37]:
```suggestion
  <ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">
```

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj [hunk 4]
```diff
@@ -22,10 +22,10 @@
 
   <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
-    <OmitResources Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</OmitResources>
+    <OmitResources Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '9.0'))">true</OmitResources>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '9.0'))">
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:28]:
```suggestion
  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">
```

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj [hunk 5]
```diff
@@ -22,10 +22,10 @@
 
   <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
-    <OmitResources Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</OmitResources>
+    <OmitResources Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '9.0'))">true</OmitResources>
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:25]:
```suggestion
    <OmitResources Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">true</OmitResources>
```

File: src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:7]:
```suggestion
  <ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">
```

File: src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj (hunk 2 above)
Comment (ViktorHofer) [reply src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:11]:
```suggestion
  <ItemGroup Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">
```

File: src/libraries/Microsoft.Bcl.Memory/tests/Microsoft.Bcl.Memory.Tests.csproj [hunk 6]
```diff
@@ -39,7 +39,7 @@
     </Compile>
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp' or $([MSBuild]::VersionLessThan('$(TargetFrameworkVersion)', '9.0'))">
```
Comment (buyaa-n) [inline src/libraries/Microsoft.Bcl.Memory/tests/Microsoft.Bcl.Memory.Tests.csproj:42]:
```suggestion
  <ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">
```

File: src/libraries/Microsoft.Bcl.Memory/tests/Microsoft.Bcl.Memory.Tests.csproj [hunk 7]
```diff
@@ -39,7 +39,7 @@
     </Compile>
   </ItemGroup>
 
-  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
+  <ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net9.0'))">
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Bcl.Memory/tests/Microsoft.Bcl.Memory.Tests.csproj:42]: nit: I don't think that we need this condition at all.

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/Microsoft.Bcl.Memory/tests/Microsoft.Bcl.Memory.Tests.csproj (hunk 7 above)
Comment (buyaa-n) [reply src/libraries/Microsoft.Bcl.Memory/tests/Microsoft.Bcl.Memory.Tests.csproj:42]: Right, good catch, it was needed when there was duplicate code, now `.NET 9` and `netstandard` shares the code
```suggestion
  <ItemGroup>
```

**Review by ViktorHofer — APPROVED** [review]

**Pull request merged**